### PR TITLE
Config management - Key-Value management support in CLI

### PR DESCRIPTION
**Pull request by tobiscr**

### CODE REVIEW

**Comment by tobiscr** [comment]
Closing PR... opening as Draft

File: cmd/config/get/bucket/cmd.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+			return Run(o)
+		},
+	}
+	return cmd
+}
+
+func Run(o *cli.Options) error {
+
+	pgConnFac := &db.PostgresConnectionFactory{
+		Host:     "localhost",
```
Comment (crabtree) [inline cmd/config/get/bucket/cmd.go:26]: This should be read from the configuration

Comment (tobiscr) [reply cmd/config/get/bucket/cmd.go]: Agree.. but I dropped that already. Per default is now SQLite used. The CLI is already using a config-file to retrieve DB credentials.

File: pkg/config/cacheentrryentity.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+func (ce *CacheEntryEntity) checksum() string {
+	if ce.Checksum == "" && ce.Data != "" {
+		ce.Checksum = ce.md5()
+	}
+	return ce.Checksum
+}
+
+func (ce *CacheEntryEntity) md5() string {
+	md5 := md5.Sum([]byte(ce.Data)) //nolint - MD5 is used for change detection, not for encryption
```
Comment (crabtree) [inline pkg/config/cacheentrryentity.go:37]: Maybe you can come up with different name than md5 for that variable? It covers the md5 import.

File: pkg/config/entity.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	}
+	if time, ok := value.(time.Time); ok {
+		return time, nil
+	}
+	return nil, fmt.Errorf("Failed to convert value '%s' (kind: %s) for field 'Created' to Time struct",
+		value, reflect.TypeOf(value).Kind())
+}
+
+func requireValidBucketName(value interface{}) (interface{}, error) {
+	bucketName := fmt.Sprintf("%s", value)
```
Comment (crabtree) [inline pkg/config/entity.go:30]: Is the empty string a valid bucket name?

File: pkg/db/marshaller.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	var err error
+	switch field.Kind() {
+	case reflect.Int64:
+		int64Value, ok := value.(int64)
+		if !ok {
+			return es.fireCastError(reflect.Int64, field, value)
+		}
+		err = field.Set(int64Value)
+	case reflect.Bool:
+		//some DBs handle booleans als integer values (0/1)
```
Comment (crabtree) [inline pkg/db/marshaller.go:94]:
```suggestion
		//some DBs handle booleans as integer values (0/1)
```

File: Makefile [hunk 5]
```diff
@@ -21,15 +21,25 @@ build: build-linux build-darwin build-linux-arm
...
+	@rm cover.out
+
+.PHONY: clean
+clean:
+	rm -rf bin
 
 .PHONY: all
-all: resolve build lint 
\ No newline at end of file
+all: resolve build test lint
```
[the remark this answers is not in the document]
Comment (tobiscr) [reply Makefile:45]: oh, right.. sorry C&P failure. Fix is coming...

File: cmd/config/config.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+			SslMode:  viper.GetBool("configManagement.db.postgres.sslMode"),
+			Debug:    o.Verbose,
+		}, nil
+	default:
+		return nil, fmt.Errorf("Database driver '%s' not supported yet", dbDriver)
+	}
+}
+
+func fileExists(file string) bool {
+	_, err := os.Stat(file)
```
[the remark this answers is not in the document]
Comment (tobiscr) [reply cmd/config/config.go:133]: Very good point! Thanks for the hint.. will enhance the check to make sure a found inode won't be a dir.

File: pkg/config/bucketmerger_test.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+			bucket = fmt.Sprintf("%s-%s", bucket, subBucket)
+		}
+		kvMap, err := loadYaml(t, fmt.Sprintf("%s.yaml", bucket))
+		if err == nil {
+			importKVMap(t, kvRepo, bucket, kvMap)
+		}
+	}
+}
+
+func fileExists(file string) bool {
```
[the remark this answers is not in the document]
Comment (tobiscr) [reply pkg/config/bucketmerger_test.go]: Omg.. you are really good! :) Redundant code is removed.

File: pkg/config/cacheentrryentity.go (hunk 2 above)
Comment (tobiscr) [reply pkg/config/cacheentrryentity.go]: Definitely, var is now renamed!

File: pkg/config/entity.go (hunk 3 above)
Comment (tobiscr) [reply pkg/config/entity.go:30]: no it isn't, but the validation should only run when the bucket string is defined. Marshaller are also executed in cases when an empty entity is provided (which is valid in cases of a delete-query, like here: https://github.com/kyma-incubator/reconciler/pull/2/files#diff-dae7efcf1d65d1b2a95798b43d9e45f920597f0889bd70a1e68fdf5a2e944909R70)

The validation for non-empty strings happens already via the annotation (see https://github.com/kyma-incubator/reconciler/pull/2/files#diff-2a76185f573c76ec3b23aff0fe2057dae5d020b1866407aadec53fd4a0fbef91R14). I'm actually abusing here the marshaller to define another validation for the value. A marshaller should normally only convert data-types from Go-Types to a format the DB understands and only fail if the value is not convertible. Here I let it fail when the provided value is not fitting to my expectations.

**Review by crabtree — DISMISSED** [review]

**Review by crabtree — DISMISSED** [review]

File: docs/configuration-management.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+      * Name of the bucket the value belongs to.
+      * Creation date of the entry.
+      * User who created the configuration value entry.
+    * When a configuration value entity is created, it will per-default be mapped to the latest version of a configuration key entity.
+    * Configuration value entities are immutable and versioned: a change of the value or any metadata leads to a new version of the configuration value entity.
+    * Before a configuration value entry can be created, the assigned validation logic (part of the the configuration key metadata) has to be executed. If the validation was successful (return value `true`), the entry gets persisted otherwise an error will be returned.
+* **Caching**
+  * A cache entry is the merge result of multiple buckets and specific to one particular cluster.
+  * It has to be possible to identify in which cache entry a specific configuration value was used.
+  * If a configuration value has changed (other attributes coupled to a configuration entry don't matter), any cache entries which were using this configuration value have to be invalidated.
```
Comment (NHingerl) [inline docs/configuration-management.md:33]:
```suggestion
  * If a configuration value has changed (other attributes coupled to a configuration entry don't matter), any cache entries that were using this configuration value must be invalidated.
```

File: docs/configuration-management.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  * The invalidation of a cache entry leads to a reconciliation for the related clusters (scheduling of a reconciliation run).
+* **CLI**
+  * A CLI is used to access the configuration management
```
Comment (NHingerl) [inline docs/configuration-management.md:36]:
```suggestion
  * A CLI is used to access the configuration management.
```

File: docs/configuration-management.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+The data of the configuration management can be distinguished in two domains:
+
+* Configuration data
+
+  Any information related to the configuration entires (metadata per configuration key, different versions of configuration entries etc.).
+
+* Caching data
+
+  A cache entity contains the merged configuration for one cluster. It also trackes which configuration entry was used in which cache entity to identify outdated cluster configurations.
```
Comment (NHingerl) [inline docs/configuration-management.md:53]:
```suggestion
  A cache entity contains the merged configuration for one cluster. It also tracks which configuration entry was used in which cache entity to identify outdated cluster configurations.
```

File: docs/configuration-management.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+|Column|Description|Data Type|Primary Key|Example|
+|--|--|--|--|--|
+|key|The configuration key|string|Yes|`my.config.key`|
+|version|Incremental version number|Integer|Yes|`1`|
+|datatype|Data type of the stored information|String|No|`integer`|
+|encrypted|Flag whether the value is plain text or encrypted|Boolean|No|`false`|
+|created|Timestamp when the entry was created|Integer|No|`123456789`|
+|user|User who created the entry|String|No|`i98765`|
+|validator|Optional logic which is executed to validated the value. Return value has to be a boolean: `true`=valid / `false`=invalid|String|No|`it >= 1 && it < 10`
```
Comment (NHingerl) [inline docs/configuration-management.md:72]:
```suggestion
|validator|Optional logic that is executed to validate the value. Return value must be a boolean: `true`=valid / `false`=invalid|String|No|`it >= 1 && it < 10`
```

File: docs/configuration-management.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+|trigger|Optional trigger functions which are executed by the reconciler|String|No|`restartPod("app=appLabel,component=abc")`
```
Comment (NHingerl) [inline docs/configuration-management.md:73]:
```suggestion
|trigger|Optional trigger functions that are executed by the reconciler|String|No|`restartPod("app=appLabel,component=abc")`
```

File: docs/configuration-management.md [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+|bucket|The bucket the configuration entry belongs to|String|Yes|`cust1`|
+|value|The value of the configuration entry|String|No|`123`|
+|created|Timestamp when the entry was created|Integer|No|`123456789`|
+|user|User who created the entry|String|No|`i98765`|
+
+#### Cache Table
+
+Requirements for the data structure layout:
+
+* Cache bucket merge result per cluster.
```
Comment (NHingerl) [inline docs/configuration-management.md:91]: Why is this one not a sentence! All the others are full sentences! :O

Comment (tobiscr) [reply docs/configuration-management.md:91]: Hm, good point. Will fix that when I'm updating the tech-docs in general. This file contains currently only some basic information.

**Review by alexandra-simeonova — APPROVED** [review]

**Review by crabtree — APPROVED** [review]

**Pull request merged**